python/pyspark/sql/tests/test_dataframe.py

```diff
@@ -547,6 +547,27 @@ def test_to_pandas_avoid_astype(self):
         self.assertEquals(types[1], np.object)
         self.assertEquals(types[2], np.float64)
 
+    @unittest.skipIf(not have_pandas, pandas_requirement_message)
+    def test_to_pandas_from_empty_dataframe(self):
+        # SPARK-29188 test that toPandas() on an empty dataframe had the correct dtypes
+        import numpy as np
+        schema = StructType([
+            StructField('double', DoubleType(), True),
+            StructField('float', FloatType(), True),
+            StructField('byte', ByteType(), True),
+            StructField('integer', IntegerType(), True),
+            StructField('long', LongType(), True),
+            StructField('short', ShortType(), True),
+        ])
```

Comment:
@dlindelof How does it work for decimal and other types? You're fixing a fundamental problem (see https://github.com/apache/spark/pull/26747#discussion_r353500751)

Can you test other type combinations, and make sure non-empty and empty types are same?

Reply:
Hi,

I've added some more types, I think we have the most important ones now. I've also checked how this behaves in the presence of nulls.

Let me know if you think I'm missing something or if I should have done something differently.